packages/route/src/Application/ResourceLocator/ResourceLocatorGenerator.php

```diff
@@ -38,13 +39,29 @@ public function generate(ResourceLocatorRequest $request): string
             $parentPath = $parentRoute?->getSlug() ?: '/';
         }
 
-        $parts = \array_map(fn ($part) => $this->pathCleanup->cleanup($part, $request->locale), $request->parts);
+        $parentPath = \rtrim($parentPath, '/');
 
-        $path = '/' . \implode('-', $parts);
+        if (null !== $request->routeSchema && null !== $this->routeSchemaProcessor) {
+            $path = $this->routeSchemaProcessor->process($request);
 
-        $parentPath = \rtrim($parentPath, '/');
+            $uniquePath = $this->createUnique(
+                $parentPath . $path,
+                $request->locale,
+                $request->webspace,
+                $request->resourceKey,
+                $request->resourceId,
+            );
+
+            if ($request->relative) {
+                return \substr($uniquePath, \strlen($parentPath));
+            }
 
-        // TODO routeSchema
+            return $uniquePath;
+        }
+
+        $parts = \array_map(fn ($part) => $this->pathCleanup->cleanup($part, $request->locale), $request->parts);
+
+        $path = '/' . \implode('-', $parts);
```

Comment:
we could create a default schema which does this, so the process is always same see above.